### PR TITLE
Support for storing logfile aggregates

### PR DESCRIPTION
**Pull request by xavierdecoster**
This PR:
* Adds support for a new -AggregatesOnly job execution mode, which enables ad-hoc retroactive parsing of log files and storing aggregates only (needed to patch the db for already imported blobs)
* Enables the stats importer (in regular execution mode) to also store total package download counts per log file to the stats db, linked to the date dimension id.

This combined will provide a means to easily query total package downloads by date.

cc @maartenba @skofman1

### CODE REVIEW

File: src/NuGet.Jobs.Common/JobRunner.cs [hunk 1]
```diff
@@ -128,6 +128,8 @@ private static void JobSetup(JobBase job, IDictionary<string, string> jobArgsDic
                 sleepDuration = 5000;
             }
 
+            job.ConsoleLogOnly = consoleLogOnly;
```
Comment (xavierdecoster) [inline src/NuGet.Jobs.Common/JobRunner.cs]: This tells the job whether it is running in console-log-only mode, which allows for conditional diagnostics configuration.

File: src/Stats.ImportAzureCdnStatistics/Job.cs [hunk 2]
```diff
@@ -45,6 +50,17 @@ public override bool Init(IDictionary<string, string> jobArgsDictionary)
                 _azureCdnAccountNumber = JobConfigurationManager.GetArgument(jobArgsDictionary, JobArgumentNames.AzureCdnAccountNumber);
                 _azureCdnPlatform = ValidateAzureCdnPlatform(azureCdnPlatform);
                 _cloudStorageContainerName = ValidateAzureContainerName(JobConfigurationManager.GetArgument(jobArgsDictionary, JobArgumentNames.AzureCdnCloudStorageContainerName));
+
+                _aggregatesOnly = JobConfigurationManager.TryGetBoolArgument(jobArgsDictionary, "AggregatesOnly");
```
Comment (maartenba) [inline src/Stats.ImportAzureCdnStatistics/Job.cs]: Make it an item in "JobArgumentNames"

File: src/Stats.ImportAzureCdnStatistics/Job.cs [hunk 3]
```diff
@@ -60,35 +76,42 @@ public override bool Init(IDictionary<string, string> jobArgsDictionary)
         {
             try
             {
-                // construct a cloud blob client for the configured storage account
```
Comment (xavierdecoster) [inline src/Stats.ImportAzureCdnStatistics/Job.cs]: This should not be re-initialized on every run. Moved to `Init` method.

File: src/Stats.ImportAzureCdnStatistics/LogFileAggregates.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    {
+        public LogFileAggregates(string logFileName)
+        {
+            LogFileName = logFileName;
+            PackageDownloadsByDate = new Dictionary<int, int>();
+        }
+
+        public string LogFileName { get; private set; }
+
+        public IDictionary<int, int> PackageDownloadsByDate { get; private set; }
```
Comment (maartenba) [inline src/Stats.ImportAzureCdnStatistics/LogFileAggregates.cs]: Any reason the date is an int and not DateTime(Offset)?

**Comment by maartenba** [comment]
A few nits, other than that LGTM

File: src/Stats.ImportAzureCdnStatistics/LogFileAggregates.cs (hunk 4 above)
Comment (xavierdecoster) [inline src/Stats.ImportAzureCdnStatistics/LogFileAggregates.cs]: Because these are not the actual dates. They are the date dimension ID's. (db normalization)

File: src/Stats.ImportAzureCdnStatistics/Job.cs (hunk 2 above)
Comment (xavierdecoster) [inline src/Stats.ImportAzureCdnStatistics/Job.cs]: Fixed

File: src/NuGet.Jobs.Common/JobBase.cs [hunk 5]
```diff
@@ -28,6 +28,8 @@ protected JobBase(EventSource jobEventSource)
 
         public JobTraceListener JobTraceListener { get; private set; }
 
+        public bool ConsoleLogOnly { get; set; }
```
Comment (skofman1) [inline src/NuGet.Jobs.Common/JobBase.cs]: why is this part of JobBase? This is relevant only for the entity creating the logger.

File: src/Stats.ImportAzureCdnStatistics/LogFileAggregates.cs (hunk 4 above)
Comment (skofman1) [inline src/Stats.ImportAzureCdnStatistics/LogFileAggregates.cs]: Can you add a comment about it?

---
In reply to: [63694833](https://github.com/NuGet/NuGet.Jobs/pull/40#discussion_r63694833) [](ancestors = 63694833)

File: src/Stats.ImportAzureCdnStatistics/LogFileProcessor.cs [hunk 6]
```diff
@@ -70,32 +74,73 @@ internal class LogFileProcessor
...
+                                var downloadsByDate =
+                                    table.AsEnumerable()
+                                        .GroupBy(e => e.Field<int>("Dimension_Date_Id"))
+                                        .Select(e => new KeyValuePair<int, int>(e.Key, e.Count()));
+
+                                foreach (var keyValuePair in downloadsByDate)
+                                {
+                                    logFileAggregates.PackageDownloadsByDate.Add(keyValuePair.Key, keyValuePair.Value);
+
+#if DEBUG
```
Comment (skofman1) [inline src/Stats.ImportAzureCdnStatistics/LogFileProcessor.cs]: Can we use log verbosity instead of a pragma?

File: src/Stats.ImportAzureCdnStatistics/LogFileProcessor.cs [hunk 7]
```diff
@@ -70,32 +74,73 @@ internal class LogFileProcessor
...
+                        await _warehouse.StoreLogFileAggregatesAsync(logFileAggregates);
+                    }
+                }
+
+                if (!aggregatesOnly)
+                {
+                    await ArchiveBlobAsync(logFile);
+                }
             }
             catch (Exception e)
```
Comment (skofman1) [inline src/Stats.ImportAzureCdnStatistics/LogFileProcessor.cs]: log the exception

**Comment by skofman1** [comment]
Added some comments

File: src/NuGet.Jobs.Common/JobBase.cs (hunk 5 above)
Comment (xavierdecoster) [inline src/NuGet.Jobs.Common/JobBase.cs]: The entity creating the logger in fact is the type inheriting from JobBase. Didn't want to expand the scope of this PR by refactoring the entire jobs code base.

Comment (skofman1) [inline src/NuGet.Jobs.Common/JobBase.cs]: But why is this relevant to all Jobs?

Comment (xavierdecoster) [inline src/NuGet.Jobs.Common/JobBase.cs]: It's the only way the JobBase implementor knows that it is running in this mode, as the -ConsoleLogOnly argument gets stripped out in the JobRunner before passing the arguments to the JobBase.Init method.

File: src/Stats.ImportAzureCdnStatistics/LogFileAggregates.cs (hunk 4 above)
Comment (xavierdecoster) [inline src/Stats.ImportAzureCdnStatistics/LogFileAggregates.cs]: Addressed with https://github.com/NuGet/NuGet.Jobs/pull/40/commits/72857e4f24d53e29a32522802852f9cd4ad38612

File: src/Stats.ImportAzureCdnStatistics/LogFileProcessor.cs (hunk 7 above)
Comment (xavierdecoster) [inline src/Stats.ImportAzureCdnStatistics/LogFileProcessor.cs]: Good catch! It is logged as part of moving the blob to the deadletter container in the blobs metadata, however that predates our improved logger :)
Fixed with https://github.com/NuGet/NuGet.Jobs/pull/40/commits/d6856623feee7103e9ea1e8da95410a4b855d980

File: src/Stats.ImportAzureCdnStatistics/LogFileProcessor.cs (hunk 6 above)
Comment (xavierdecoster) [inline src/Stats.ImportAzureCdnStatistics/LogFileProcessor.cs]: Fixed with https://github.com/NuGet/NuGet.Jobs/pull/40/commits/d6856623feee7103e9ea1e8da95410a4b855d980

**Comment by skofman1** [comment]
:ship:
